compiler/src/dotty/tools/dotc/transform/SymUtils.scala

```diff
@@ -259,6 +259,13 @@ object SymUtils:
         && self.owner.linkedClass.is(Case)
         && self.owner.linkedClass.isDeclaredInfix
 
+    /** Is symbol declared experimental? */
+    def isExperimental(using Context): Boolean =
+      (self eq defn.ExperimentalAnnot)
+      || self.hasAnnotation(defn.ExperimentalAnnot)
+      || self.allOverriddenSymbols.nonEmpty && self.allOverriddenSymbols.forall(_.hasAnnotation(defn.ExperimentalAnnot)) // TODO infer @experimental?
```

Comment:
I would say a method is experimental if at least one method it overrides is experimental:
```suggestion
      || self.allOverriddenSymbols.exists(_.hasAnnotation(defn.ExperimentalAnnot)) // TODO infer @experimental?
```
but I think more properly when we do override checks in RefChecks, we should disallow overriding an experimental method from a non-experimental method, same with subclassing an experimental class.

Reply:
For methods, I am not sure what we should do with the following case

```scala
trait A:
  @experimental def f: Int
trait B:
  def f: Int
trait C extends A, B:
  def f: Int = 3
```
Is `C.f` experimental? It implements an experimental API but at the same time it implements a stable API. If we look at it as implementing `B.f` it can be considered as non-experimental.

Reply:
It implements both and if A.f is removed that might have binary compatibilty implications  on C.f and its users, so yes.

Reply:
Applying the LSP, the only thing that should be forbidden is when an experimental method implements/overrides a non-experimental one, for example:
```scala
trait A:
  def f: Int
  def g: Int = 3
trait B:
  @experimental def f: Int = 4 // error
  @experimental override def g: Int = 5 // error
```
It's perfectly fine to implement or override an experimental method with a non-experimental one.

Reply:
> It's perfectly fine to implement or override an experimental method with a non-experimental one.

I'm not convinced, it's at least not source-compatible since if that method gets removed, your usage of `override` is now an error. I would err on the side of caution and disallow both directions.

Reply:
> It implements both and if A.f is removed that might have binary compatibilty implications on C.f and its users, so yes.

No, it can't. Users of `C.f` will always produce bytecode that call `C.f` itself, never `A.f` or any of the bridges generated in `C` to adapt it to `A.f`.

Reply:
> but I think more properly when we do override checks in RefChecks, we should disallow overriding an experimental method from a non-experimental method

Do you mean that we need to annotate each experimental method with `@experimental`? That would imply that all members of an experimental class must be individually annotated.

Reply:
If the class itself is experimental then when you subclass it, the subclass should be marked experimental, but no need to mark the individual methods as experimental

Reply:
> > It's perfectly fine to implement or override an experimental method with a non-experimental one.
> 
> I'm not convinced, it's at least not source-compatible since if that method gets removed, your usage of `override` is now an error. I would err on the side of caution and disallow both directions.

*Any* change of public/protected API, even experimental, is always going to be source incompatible. Removing an experimental top-level class could change how some name resolution behaves, even if it did not actually resolve to that class before. You can't protect people from theoretical source incompatibilities. Don't try.